lerobot/common/policies/vqbet/modeling_vqbet.py

```diff
@@ -0,0 +1,2382 @@
+            state_rep_flat = state_rep.view(state_rep.size(0), -1, state_rep.size(1))
+            state_rep_flat, vq_code, vq_loss_state = self.vq_layer(state_rep_flat)
+            state_vq = state_rep_flat.view(*state_rep_shape, -1)
+            vq_code = vq_code.view(*state_rep_shape, -1)
+            vq_loss_state = torch.sum(vq_loss_state)
+            if required_recon:
+                recon_state = self.decoder(state_vq)
+                recon_state_ae = self.decoder(state_rep)
+                if self.config.n_action_pred_chunk == 1:
+                    return state_vq, vq_code, recon_state, recon_state_ae
+                else:
+                    return (
+                        state_vq,
+                        vq_code,
+                        torch.swapaxes(recon_state, -2, -1),
+                        torch.swapaxes(recon_state_ae, -2, -1),
+                    )
+            else:
+                return state_vq, vq_code
+
+    def vqvae_forward(self, state):
+        state = self.preprocess(state)
+        state_rep = self.encoder(state)
+        state_rep_shape = state_rep.shape[:-1]
+        state_rep_flat = state_rep.view(state_rep.size(0), -1, state_rep.size(1))
+        state_rep_flat, vq_code, vq_loss_state = self.vq_layer(state_rep_flat)
+        state_vq = state_rep_flat.view(*state_rep_shape, -1)
+        vq_code = vq_code.view(*state_rep_shape, -1)
+        vq_loss_state = torch.sum(vq_loss_state)
+
+        dec_out = self.decoder(state_vq)
+        encoder_loss = (state - dec_out).abs().mean()
+
+        rep_loss = encoder_loss + vq_loss_state * 5
+
+        metric = (
+            encoder_loss.clone().detach(),
+            vq_loss_state.clone().detach(),
+            vq_code,
+            rep_loss.item(),
+        )
+        return rep_loss, metric
+
+
+    def load_state_dict(self, *args, **kwargs):
+        super(VqVae, self).state_dict(self, *args, **kwargs)
+        self.eval()
+        self.discretized = True
+
+
+
+
+
+def pretrain_vqvae(vqvae_model, discretize_step, actions):
+    if vqvae_model.config.n_action_pred_chunk == 1:
+        # not using action chunk
+        actions = actions.reshape(-1, 1, actions.shape[-1])
+    else:
+        # using action chunk
+        slices = []
+        slices.extend([actions[:, j:j+vqvae_model.config.n_action_pred_chunk, :] for j in range(actions.shape[1]+1-vqvae_model.config.n_action_pred_chunk)])
+        actions = torch.cat(slices, dim=0)
+
+
+    actions = actions.to(get_device_from_parameters(vqvae_model))
+
+    loss, metric = vqvae_model.vqvae_forward(
+        actions
+    )
+    n_different_codes = len(torch.unique(metric[2]))
+    n_different_combinations = len(torch.unique(metric[2], dim=0))
+    vqvae_model.optimized_steps += 1
+    if vqvae_model.optimized_steps >= discretize_step:
+        vqvae_model.discretized = True
+    return loss, n_different_codes, n_different_combinations
+
+
+def exists(val):
```

Comment:
This appears to be defined twice. Also, is it okay if we avoid using these simple functions that wrap short 1-liners? It adds a layer of indirection which would make it harder to read the code (reader has to go and check what `exists` does).

Reply:
deleted the parts that were defined twice and some short functions. (`def default`, `def exists`, `def round_up_multiple`, and `def l2norm`) But I left some of the functions that is used in the very low-level part of RVQ since removing all these short function in low-level of RVQ doesn't necessarily lead to more readable code.

Reply:
Excellent. Thank you :)